org.jacoco.core.test/src-java7/org/jacoco/core/test/filter/TryWithResourcesTest.java

```diff
@@ -0,0 +1,175 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2017 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.filter;
+
+import java.io.IOException;
+
+import org.jacoco.core.analysis.ICounter;
+import org.jacoco.core.internal.Java9Support;
+import org.jacoco.core.test.TargetLoader;
+import org.jacoco.core.test.validation.ValidationTestBase;
+import org.jacoco.core.test.filter.targets.TryWithResources;
+import org.junit.Test;
+
+/**
+ * Test of filtering of a bytecode that is generated for a try-with-resources
+ * statement.
+ */
+public class TryWithResourcesTest extends ValidationTestBase {
+
+	public TryWithResourcesTest() {
+		super("src-java7", TryWithResources.class);
+	}
+
+	/**
+	 * {@link TryWithResources#test()}
+	 */
+	@Test
+	public void test() {
+		assertLine("test.before", ICounter.FULLY_COVERED);
+		// without filter next line covered partly:
+		assertLine("test.try", ICounter.FULLY_COVERED);
+		assertLine("test.open1", ICounter.FULLY_COVERED);
+		assertLine("test.open2", ICounter.FULLY_COVERED);
+		assertLine("test.open3", ICounter.FULLY_COVERED);
+		assertLine("test.body", ICounter.FULLY_COVERED);
+		// without filter next line has branches:
+		assertLine("test.close", ICounter.EMPTY);
+		assertLine("test.catch", ICounter.NOT_COVERED);
+		assertLine("test.finally", ICounter.PARTLY_COVERED);
+	}
+
+	/**
+	 * {@link TryWithResources#test2()}
+	 */
+	@Test
+	public void test2() {
+		assertLine("test2.before", ICounter.FULLY_COVERED);
+		// without filter next line covered partly:
+		assertLine("test2.try", ICounter.FULLY_COVERED);
+		assertLine("test2.open1", ICounter.FULLY_COVERED);
+		assertLine("test2.open2", ICounter.FULLY_COVERED);
+		assertLine("test2.open3", ICounter.FULLY_COVERED);
+		assertLine("test2.body", ICounter.FULLY_COVERED);
+		// without filter next line has branches:
+		assertLine("test2.close", ICounter.EMPTY);
+		assertLine("test2.catch", ICounter.NOT_COVERED);
+		assertLine("test2.finally", ICounter.PARTLY_COVERED);
+		assertLine("test2.after", ICounter.FULLY_COVERED);
+	}
+
+	/**
+	 * {@link TryWithResources#returnInBody()}
+	 */
+	@Test
+	public void returnInBody() {
+		// without filter next line covered partly:
+		assertLine("returnInBody.try", ICounter.FULLY_COVERED);
+		assertLine("returnInBody.open", ICounter.FULLY_COVERED);
+		// without filter next line has branches:
+		// TODO with javac 1.8.0_31 part of return instructions belong to close
+		// assertLine("returnInBody.close", ICounter.EMPTY);
```

Comment:
@marchof Excuse me for the delay with reply - was at DevoxFR, where BTW session "Java Code Coverage Mechics" was a quite success.

On a subject:

@bjkail `java.version` maybe not so fragile.

> ASM to detect the missing line number

is fragile in terms that it allows compiler to regress on this.

@marchof 

> Maybe our integration tests for filtering are too specific.

I personally like such specificity for the following reasons
* not all examples here reach full coverage (no uncovered lines and branches) and also IMO will not reach in case of future filters
* consistency across tests
* allows to see exactly what user will see
* allows to catch regressions in compilers

> We already have plenty of validation test for expected covered/missed lines

Apparently not for such tricky control flows as shown here :wink: 

Do you see any troubles to simply bump version of javac 8 up to latest update as in my initial proposal - "past is past"?